### ISSUE TITLE
Pretraining hyperparameters

### ISSUE THREAD

**Issue by gabrielhuang** [Sha-Lab/FEAT#29]
Hi there,

Are the default hyperparameters in pretrain.py good or are there variations depending on the architecture or the dataset ?

Thanks

**Comment by Han-Jia** [comment]
The parameters depend on the architecture. 

For ConvNet, the initial learning rate should be around 0.01, and for ResNet the initial learning rate is 0.1.

**Comment by smallbox120** [comment]
I use lr 0.1 and pretrain ResNet12, the pre-trained initialization obtained doesn't achieve performance close to the paper(but the download initializations can). Actually, the performance gap is up to 4% in MiniImageNet. Can you please provide the script of pretraining that can reproduce the results. It's somewhat shocking, that different pretraining hyperparameters can result in such huge performance gap.

**Comment by Han-Jia** [comment]
When pre-training on ResNet12, the model should be optimized with lr=0.1 for long epochs. During our empirical study, training with lr=0.1 for 200-300 epochs and then drop the learning rate generalizes better.

**Comment by SnowyJune973** [comment]
> When pre-training on ResNet12, the model should be optimized with lr=0.1 for long epochs. During our empirical study, training with lr=0.1 for 200-300 epochs and then drop the learning rate generalizes better.

Thanks. Besides, what's the batch size you used when pretraining TieredImageNet+ResNet12? Looks like bigger batch sizes are needed to make the model coverge quicker. (At present I'm using batch_size=2048 on 4 V100 GPUs, but it still costs 29h to finish all 500 epochs of pretraining)

**Comment by Han-Jia** [comment]
Due to the GPU memory limit, I use the batch-size128.

It indeed costs a long time during the pre-training stage. Actually, we do not need to validate the model after each epoch, we usually obtain a good model after 300 epochs.

**Comment by SnowyJune973** [comment]
> Due to the GPU memory limit, I use the batch-size128.
> 
> It indeed costs a long time during the pre-training stage. Actually, we do not need to validate the model after each epoch, we usually obtain a good model after 300 epochs.

Thanks a lot !

**Comment by MinkiJ** [comment]
Hi, thank you for your detailed experimental code.

I could reproduce your work with the provided pre-trained backbones. However, I got lower performances when I employed my own pre-trained backbone.

Could you provide detailed hyperparameters to reproduce the backbone networks?

After I read several issues about pre-training, I set
lr: 0.1
batch_size: 128
max_epoch: 500
schedule: [200, 350]
gamma: 0.1
and set the other parameters to default

to train the ResNet18 backbone on the mini-ImageNet dataset, but it results worse than yours.

Thank you!

**Comment by Han-Jia** [comment]
Hi,

For MiniImageNet, I use this:

lr = 0.1, batch-size = 128, max_epoch = 500, schedule: [350, 400, 440, 460, 480], gamma: 0.1

In addition, during the experiments, I find ResNet18 could not perform as well as the ResNet12. It is reasonable that ResNet18 gets a lower validation accuracy during pre-training. The performance of ResNet18 may depend on the detailed architecture, e.g., remove which downsampling layers.

**Comment by smallbox120** [comment]
> Hi,
> 
> For MiniImageNet, I use this:
> 
> lr = 0.1, batch-size = 128, max_epoch = 500, schedule: [350, 400, 440, 460, 480], gamma: 0.1
> 
> In addition, during the experiments, I find ResNet18 could not perform as well as the ResNet12. It is reasonable that ResNet18 gets a lower validation accuracy during pre-training. The performance of ResNet18 may depend on the detailed architecture, e.g., remove which downsampling layers.

I did exactly that, still can't get the result reported in the paper, which is about 2% behind on 5-way 1-shot (64.5%). This is crucial if we want to do further analysis on this task.

**Comment by Han-Jia** [comment]
Hi, 

I run the command again and get almost the same results.

What is the validation accuracy with the pre-trained model? The 16-way 1-shot  classification accuracy on the meta-val set should be around 46%.

**Comment by smallbox120** [comment]
> Hi,
> 
> I run the command again and get almost the same results.
> 
> What is the validation accuracy with the pre-trained model? The 16-way 1-shot classification accuracy on the meta-val set should be around 46%.

Actually, no. the final validation accuracy of the pretraining reaches around 28% (acc_dist) and 32% (acc_sim). The pretraining script I use is:

python pretrain.py --batch_size 128 --max_epoch 500 --lr 0.1 --dataset MiniImageNet --backbone_class Res12 --schedule  350 400 440 460 480 --gamma 0.1

and the meta-training script (copied from README, execpt the path for init_weights and designating gpu) I use is:

python train_fsl.py  --max_epoch 200 --model_class FEAT --use_euclidean --backbone_class ConvNet --dataset MiniImageNet --way 5 --eval_way 5 --shot 1 --eval_shot 1 --query 15 --eval_query 15 --balance 1 --temperature 64 --temperature2 16 --lr 0.0001 --lr_mul 10 --lr_scheduler step --step_size 20 --gamma 0.5 --gpu 8 --init_weights ./MiniImageNet-Res12-Pre/0.1_0.1_\[350\,\ 400\,\ 440\,\ 460\,\ 480\]/max_acc_dist.pth --eval_interval 1 --gpu 1

Images are 224X224 pre-augmentation, and croped to 84X84 after.

**Comment by Han-Jia** [comment]
I use the pretrain command above and the code in the repo, and it outputs during training:
[Dist] best epoch 423, current best val acc=0.4216
[Sim] best epoch 423, current best val acc=0.4665
and
epoch 500, train 1/300, loss=0.3703 acc=0.8984
epoch 500, train 101/300, loss=0.3586 acc=0.9219
epoch 500, train 201/300, loss=0.1957 acc=0.9297
[Dist] best epoch 423, current best val acc=0.4216
[Sim] best epoch 423, current best val acc=0.4665
at last.

during the pre-training, please also check:
1. how about the training loss, does it decrease to nearly 0.2 after 400 epochs.
2. the random crop from the raw images are used during training, and the center crop is used during inference.

**Comment by smallbox120** [comment]
> I use the pretrain command above and the code in the repo, and it outputs during training:
> [Dist] best epoch 423, current best val acc=0.4216
> [Sim] best epoch 423, current best val acc=0.4665
> and
> epoch 500, train 1/300, loss=0.3703 acc=0.8984
> epoch 500, train 101/300, loss=0.3586 acc=0.9219
> epoch 500, train 201/300, loss=0.1957 acc=0.9297
> [Dist] best epoch 423, current best val acc=0.4216
> [Sim] best epoch 423, current best val acc=0.4665
> at last.
> 
> during the pre-training, please also check:
> 
> 1. how about the training loss, does it decrease to nearly 0.2 after 400 epochs.
> 2. the random crop from the raw images are used during training, and the center crop is used during inference.

1. The training loss is near 0.2, but the loss seems to converge after 100 epochs, even though the best validation accuracy is achieved somewhere around epoch 425. 
2. Data augmentation is used since I run the exact same code cloned from this repository.

**Comment by smallbox120** [comment]
Another strange thing is that the validation accuracy sees a trend of slowly descending before epoch 350, then after the learning rate changes, the accuracy starts to improve and reaches  best performance around epoch 420, I wonder is this normal?

Meanwhile, I wonder which version of pytorch does the auther use?

**Comment by Han-Jia** [comment]
Hi,

We observed a similar phenomenon during our experiments. The validation accuracy changes a lot after the first learning rate decay. The PyTorch 1.4.0 is used.

**Comment by RobootMan** [comment]
> Hi,
> 
> I run the command again and get almost the same results.
> 
> What is the validation accuracy with the pre-trained model? The 16-way 1-shot classification accuracy on the meta-val set should be around 46%.

I run the command in this issue and code in this repo to pretrain the Res12, but the performance is worse than the Res12-pre.pth ( you provided in this repo ), so  I wonder if you could share me the hyperparameters you used while pretraining Res12?

**Comment by Han-Jia** [comment]
Hi，

The command is :
python3.7 pretrain.py --lr 0.1 --batch_size 128 --max_epoch 500 --backbone_class Res12 --schedule 350 400 440 460 480 --ngpu 1 --gamma 0.1 

I use the MiniImageNet dataset from [this link](https://drive.google.com/file/d/1BCxmqLANXHbBaWs8A7_jqfVUv8mydp5R/view?usp=sharing).
